rfcs/12-primitive-shapes.md

```diff
@@ -0,0 +1,514 @@
+```
+The base primitive types only define the shape (`Circle2d`) and size (`radius`) of the geometry about the origin. Once generated, the mesh can have a transform applied to it like any other mesh, and it is no longer tied to the primitive that generated it. The `Default::default()` implementation of primitives should be a "unit" variant, such as a circle with diameter 1.0, or a box with all edges of length 1.0.
+
+Meshing could be naturally extended with other libraries or parameters. For example, a sphere by default might use `Icosphere`, but could be extended to generate UV spheres or quad spheres. For 2d types, we can use a crate such as `lyon` to generate 2d meshes from parameterized primitive shapes within the `Meshable` interface.
+
+### Bounding
+
+For use in spatial queries, broad phase collision detection, and raycasting, `Bounding` and `Bounding2d` traits are provided. These traits are implemented for types that define position in addition to the shape and size defined in the base primitive. The basic functionality of this trait is to check whether one bounding shape is contained within another bounding shape.
+
+### Colliders
+
+Colliders can provide intersection checks against other colliders, as well as check if they are within a bounding volume.
+
+### 3D and 2D
+
+This RFC provides independent 2d and 3d primitives. Recall that the purpose of this is to provide lightweight types, so there are what appear to be duplicates in 2d and 3d, such as `Line` and `Line2d`. Note that the 2d version of a line is smaller than its 3d counterpart because it is only defined in 2d. 3d geometry (or 2d with depth) is assumed to be the default for most cases. The names of the types were chosen with this in mind.
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
+struct Angle(f32);
+```
+### Traits
+
+These traits are provided as reference to illustrate how these might be used. The details of implementation and interface should be determined in a separate RFC, PR, or independent prototypes.
+
+```rust
+trait Meshable{
+  fn mesh(&self) -> Mesh;
+};
+
+trait Bounding {
+  fn within(&self, other: &impl Bounding) -> bool;
+  fn contains(&self, collider: &impl Collider) -> bool;
+}
+
+trait Bounding2d {
+  fn within(&self, other: &impl Bounding2d) -> bool;
+  fn contains(&self, collider: &impl Collider2d) -> bool;
+}
+
+trait Collider {
+  fn collide(&self, other: &impl Collider) -> Option(Collision);
+  fn within(&self, bounds: &impl Bounding) -> bool;
+}
+
+trait Collider2d {
+  fn collide(&self, other: &impl Collider2d) -> Option(Collision2d);
+  fn within(&self, bounds: &impl Bounding2d) -> bool;
+}
+```
+
+### 3D Geometry Types
+
+```rust
+struct Point(Vec3)
+
+/// Vector direction in 3D space that is guaranteed to be normalized through its getter/setter.
+struct Direction(Vec3)
```

Comment:
I feel like this could be expensive in practice due to unnecessary normalization happening. The comment mentions getters and setters, would there be other methods? It might be worth fleshing this out as part of the RFC to clarify. For example would this type support ops like Add or Sub? Would the result of such operations be `Vec3`?

Reply:
Agreed here.
Though optionally we could create some wrapper like a `NormalizedVec3` which guarantees normalization.
But yes, it would be nice to see this API fleshed out.

Reply:
There is some discussion [here](https://github.com/bevyengine/rfcs/pull/12#issuecomment-848206301) (below), as well as a proposal to use the typestate pattern to normalize and memoize on access only.